New: Add rule to check `Set-Cookie` header

qzhou1607 opened the pull request:
Fix #24 

Note: It's possible for one response to contain multiple `Set-Cookie` headers. The `cdp` connector concatenate all `Set-Cookie` headers into one single string (no delimiter to split it into an array). And `jsdom` returns an array containing all the `Set-Cookie` headers. So analyzing the `Set-Cookie` headers one by one seems impossible using the `cdp` connector.

The first commit is cherry picked from #370. So only commit 2 needs to be reviewed in this PR.

sarvaje commented on src/lib/rules/validate-set-cookie-header/validate-set-cookie-header.ts:
```diff
@@ -0,0 +1,226 @@
+                return;
+            }
+
+            setCookieHeaders.forEach(async (setCookieValue) => {
```
this method is too big. I think you can refactor it and move some of the checks to functions. i.e: `checkDateFormat` or `validateCookiePrefix`

qzhou1607 commented on src/lib/rules/validate-set-cookie-header/validate-set-cookie-header.ts:
```diff
@@ -0,0 +1,226 @@
+/**
```
@sarvaje Done. I refactored the validate function as well and added more tests.

sarvaje replied:
There are still some variables without a type

qzhou1607 replied:
@sarvaje just added type annotation to all the variables I can find. One question - I can't find much information about how to add type annotation to variables from array destructuring assignment like:
```
const [nameValuePair, ...directivePairs] = setCookieValue.split(';');
```
And if I change that to
```
let nameValuePair: string;
let directivePairs: Array<string>;

[nameValuePair, ...directivePairs] = setCookieValue.split(';');
```
It wouldn't pass the lint. Do you know how to add the type in this case?

sarvaje replied:
did you try this?
`const [nameValuePair, ...directivePairs]: Array<string> = setCookieValue.split(';');`

qzhou1607 replied:
@sarvaje That worked :).

alrra commented on .sonarrc:
```diff
@@ -25,6 +25,6 @@
         "no-protocol-relative-urls": "warning",
         "ssllabs": "off",
         "x-content-type-options": "warning",
-        "strict-transport-security": "warning"
```
`strict-transport-security` has been removed.

alrra commented on docs/user-guide/rules/index.md:
```diff
@@ -34,4 +34,4 @@
 * [`no-protocol-relative-urls`](no-protocol-relative-urls.md)
 * [`ssllabs`](ssllabs.md)
 * [`x-content-type-options`](x-content-type-options.md)
-* [`strict-transport-security`](strict-transport-security.md)
```
`strict-transport-security` has been removed.

alrra commented on docs/user-guide/rules/validate-set-cookie-header.md:
```diff
@@ -0,0 +1,179 @@
+
+This rule validates the `set-cookie` header and confirms that the
+`Secure` and `HttpOnly` directives are defined when sent from a
+secure origin (HTTPS).
```
Maybe we should rename this rule to something else since we are only looking for some specific directives. 

Nothing comes to mind, but maybe something in line lines of `secure-cookies` or `no-insecure-cookies`?

molant replied:
I think this comment is obsolete as we are now validating other directives. I'm ok with `validate-set-cookie-header`.

alrra commented on docs/user-guide/rules/validate-set-cookie-header.md:
````diff
@@ -0,0 +1,179 @@
+header is as follows:
+
+```text
+Set-Cookie: <name>=<value>[; <Max-Age>=<age>] [; expires=<date>][; domain=<domain_name>] [; path=<some_path>][; secure][; HttpOnly]
````
This is to complex, and personally, I don't think it should be included.

alrra commented on docs/user-guide/rules/validate-set-cookie-header.md:
Drop `the` => `if sites are secure`

alrra commented on docs/user-guide/rules/validate-set-cookie-header.md:
`__Host-`

alrra commented on src/lib/rules/validate-set-cookie-header/validate-set-cookie-header.ts:
```diff
@@ -0,0 +1,307 @@
+        /** A collection of accepted attributes */
+        const acceptedCookieAttributes: Array<string> = ['expires', 'max-age', 'domain', 'path', 'secure', 'httponly', 'samesite', 'priority'];
+        // See https://stackoverflow.com/questions/19792038/what-does-priority-high-mean-in-the-set-cookie-header for details about the `priority` attribute.
+        /** A collection of illegal characters in cookie name */
```
Can you provide a link here?

alrra commented on src/lib/rules/validate-set-cookie-header/validate-set-cookie-header.ts:
```diff
@@ -0,0 +1,307 @@
+        // See https://stackoverflow.com/questions/19792038/what-does-priority-high-mean-in-the-set-cookie-header for details about the `priority` attribute.
+        /** A collection of illegal characters in cookie name */
+        const illegalCookieNameChars: string = '()<>@,;:\"/[]?={}'; // eslint-disable-line no-useless-escape
+        /** A collection of illegal characters in cookie value */
```
Can you provide a link here?

alrra commented on src/lib/rules/validate-set-cookie-header/validate-set-cookie-header.ts:
```diff
@@ -0,0 +1,307 @@
+        const validateMaxAgeCompatibility = (parsedSetCookie: ParsedSetCookie): ValidationErrors => {
+            const cookieName: string = parsedSetCookie.name;
+            const errors: ValidationErrors = [];
+            const maxAgeCompatibilityError: string = `Internet Explorer (ie 6, ie 7, and ie8) doesn't support 'max-age' directive in the '${headerName}' header to set '${cookieName}'.`;
```
Use `IE` instead of `ie`.

alrra commented on tests/lib/rules/validate-set-cookie-header/tests.ts:
```diff
@@ -0,0 +1,133 @@
+
+
+// Error messages.
+const messages = (cookieName: string = 'cookiename'): { [key: string]: string } => {
```
Why not just use:

```js
const messages = { 
    hasDomainHostPrefixError: '...', 
    ...
}
```

and then `messages.hasDomainHostPrefixError`.

Or directly the message if the message doesn't repeat itself.

qzhou1607 replied:
@alrra This is because `cookieName` can be different for a certain message, and it seems redundant to repeat itself when `cookieName` is the only difference. So the function is needed to interpolate the `cookieName`.

alrra commented:
> Note: It's possible for one response to contain multiple Set-Cookie headers. The cdp connector concatenate all Set-Cookie headers into one single string (no delimiter to split it into an array). 

Can you provide an example of the output. Thanks!

qzhou1607 commented:
Just found the delimiter to split the `set-cookie` header string in `cdp` when pasting the example ... Added `cdp` back.
>@qzhou1607 Can you provide an example of the output. Thanks!

molant commented:
@qzhou1607 there are some tests failing. Can you please take a look?

qzhou1607 commented:
@alrra I can't reproduce these failed tests using `cdp` in local https://travis-ci.org/sonarwhal/sonar/jobs/262425172#L2052. Is there a way to debug in travis? Also, I see a `cdp.lock.STALE` file created sometimes when using `cdp`, should I commit that file as well?

molant commented:
> I see a cdp.lock.STALE file created sometimes when using cdp, should I commit that file as well?

No, we should add thst extension to .gitignore if it isn't there already.

alrra commented:
> >  I see a cdp.lock.STALE file created sometimes when using cdp, should I commit that file as well?
> No, we should add thst extension to .gitignore if it isn't there already.

Done in https://github.com/sonarwhal/sonar/commit/12646fc9d3b9f16c279ff13f4d7636b7a2ee0109.

alrra commented:
> I can't reproduce these failed tests using cdp in local https://travis-ci.org/sonarwhal/sonar/jobs/262425172#L2052.

From just pushing a new build with some console.logs, it seems the `set-cookie` header is not present in [`response.headers`](https://github.com/qzhou1607/Sonar/blob/7c04a2c8e99e013f67fd2c293fe993bb0d60417e/src/lib/rules/no-insecure-cookies/no-insecure-cookies.ts#L255).

>  Is there a way to debug in travis?

You can [use Docker, and install a Travis CI image locally](https://docs.travis-ci.com/user/common-build-problems/#Troubleshooting-Locally-in-a-Docker-Image).

sarvaje commented:
@qzhou1607 @alrra it looks like the problem is the param `--headless` running chrome. If you remove that param it should work.

molant commented:
@qzhou1607 can you update your latest commit to link to @sarvaje's comment so we know why we remove it?

qzhou1607 commented:
@molant Link to the comment was added to the commit message.

alrra commented on src/lib/connectors/cdp/cdp-launcher.ts:
```diff
@@ -31,7 +31,9 @@ export class CDPLauncher extends Launcher {
         if (isCI) {
-            chromeFlags.push('--headless', '--disable-gpu');
+            // Running in `headless` mode is causing error in tests for `set-cookie` header.
+            // So the `--headless` flag was removed.
```
Using `--headless`  solved the problems we were having with Travis CI (see: #183).

alrra replied:
> and `--headless` breaks the tests for the cookies rules...

Removing it [will break the test for quite a few other rules](https://travis-ci.org/sonarwhal/sonar/jobs/263597186). :(

Maybe there is another way?

qzhou1607 replied:
And the `--headless` mode only ignores the `set-cookie` header, once I switched the header name to something else in the test,  like `strict-transport-security`,  the configured header is recognized in cdp again :(.

sarvaje replied:
@samccone do you have any idea why this happen? It looks like when we use the flag `--headless` the header `set-cookie` disappear in the event `onResponseReceive`, isn't it @qzhou1607?
Thanks!

ebidel replied:
This looks like https://bugs.chromium.org/p/chromium/issues/detail?id=692090. It'll take a bit to get to stable.

alrra replied:
Thanks again, @ebidel!

molant commented:
@qzhou1607 @alrra @sarvaje this is what we think we should do:

* Fix conflicts and update the tests to use [`skip`](https://github.com/avajs/ava#testskiptitle-implementation)
* Open new issue to update tests once https://bugs.chromium.org/p/chromium/issues/detail?id=692090 is in stable
* Merge current PR

Thoughts?

qzhou1607 commented:
@molant Just to make sure: This means only skipping the tests in `set-cookie` rule when using `cdp`, right? The tests do work for `jsdom`.
>* Fix conflicts and update the tests to use [`skip`](https://github.com/avajs/ava#testskiptitle-implementation)

molant commented:
> This means only skipping the tests in set-cookie rule when using cdp, right? 
Yes. If jsdom is not an issue then just not run the tests in CDP.
Also we should add back --headless if we haven't already.

qzhou1607 commented:
@alrra @molant Done.

sarvaje commented on src/lib/rules/no-insecure-cookies/no-insecure-cookies.ts:
```diff
@@ -0,0 +1,314 @@
+        const headerName: string = 'set-cookie';
+
+        /** Object containing set-cookie header information after parsing */
+        type ParsedSetCookie = { // eslint-disable-line no-unused-vars
```
Maybe should we move this to the folder `types`?

molant replied:
If these aren't going to be used anywhere else maybe a types file in the same folder as the rule? Don't know :/

molant commented on src/lib/rules/no-insecure-cookies/no-insecure-cookies.ts:
```diff
@@ -0,0 +1,314 @@
+                return await Promise.all(promises);
+            };
+
+            for (const setCookeHeader of setCookieHeaders) {
```
`setCookieHeader`, there's a missing `i`

molant commented on src/lib/rules/no-insecure-cookies/no-insecure-cookies.ts:
```diff
@@ -0,0 +1,314 @@
+
+        // Validate browser compatibility of `Max-Age` .
+        const validateMaxAgeCompatibility = (parsedSetCookie: ParsedSetCookie): ValidationErrors => {
+            const cookieName: string = parsedSetCookie.name;
```
As we discussed over [gitter](https://gitter.im/sonarwhal/Lobby) let's make this a warning.
I don't think it is fair to report this as an error if the dev is doing the right thing but a browser in its support matrix doesn't implement the feature and it won't break. 

@sonarwhal/contributors let me know if you disagree!

molant replied:
I'm going through some of the documentation and I've changed my mind.

What I think we should check depending on the user's browser support matrix is:

* Only IE6-IE8:
  * If `expires` exists: OK
  * If `max-age` exists,  there should be a `expires` too
  * If `max-age` and `expires` don't exist: OK
* Only browsers different than IE6-IE8:
  * If `expires` exists: OK
  * If `max-age` exists: OK
  * If `max-age` and `expires` exists, report an issue (`max-age` takes precedence)
* All browsers:
  * If `expires` exists: OK
  * If `max-age` exists, there should be a `expires`

qzhou1607 replied:
@molant What should happen if the browser is different than IE6-IE8, like chrome, but ONLY `max-age` exists? According to the second scenario in the list ('only browsers different than IE6-IE8'), it should pass (If `max-age` exists: OK); But according to the third scenario ('All browsers') in the list,it should throw an error/warning because there should be a `expires`?

molant replied:
That depends on the list of supported browsers the user has configured. If the user only cares about chrome 45+, then your example is valid. If the configuration says chrome 45 and IE7, then there should be a `expires` as well.

qzhou1607 replied:
Ahh I understand now. Thank you very much for the details!

qzhou1607 commented:
@molant I modified the way to validate `max-age` and `expires` based on our previous discussion. More tests were also added to verify the rule in various browser support matrix settings.

molant commented on src/lib/rules/no-insecure-cookies/no-insecure-cookies.ts:
```diff
@@ -0,0 +1,317 @@
+            const cookieName: string = parsedSetCookie.name;
+            const errors: ValidationMessages = [];
+            const maxAgeCompatibilityMessage: string = `Internet Explorer (IE 6, IE 7, and IE 8) doesn't support 'max-age' directive in the '${headerName}' header to set '${cookieName}'.`;
+            const maxAgeAndExpireDuplicateMessage: string = `The 'max-age' attribute takes precedence when both 'expires' and 'max-age' exits.`;
```
> exist.

molant commented on docs/user-guide/rules/no-insecure-cookies.md:
```diff
@@ -0,0 +1,174 @@
+# Validate `Set-Cookie` Header (`validate-set-cookie-header`)
```
Everywhere else the rule is called `no-insecure-cookies`. This rule has added more features than the insecure ones so I agree that `validate-set-cookie-header` is a better name but it should be changed everywhere (`.sonarrc`, `docs/user-guide/rules/index.md`, and rule file and tests).

molant commented on src/lib/rules/no-insecure-cookies/no-insecure-cookies.ts:
```diff
@@ -0,0 +1,317 @@
+                return errors;
+            }
+
+            // When targetBrowsers only contain modern browsers:
```
> contains

molant commented on src/lib/rules/no-insecure-cookies/no-insecure-cookies.ts:
```diff
@@ -0,0 +1,317 @@
+        };
+
+        const loadRuleConfigs = () => {
+            // When `targetedBrowsers`is empty, all browsers are supported.
```
missing space after `targetedBrowsers`

qzhou1607 commented:
@molant Done.

molant commented on src/lib/rules/validate-set-cookie-header/validate-set-cookie-header.ts:
```diff
@@ -0,0 +1,317 @@
+
+        const loadRuleConfigs = () => {
+            // When `targetedBrowsers` is empty, all browsers are supported.
+            supportOlderBrowsers = context.targetedBrowsers.length === 0 || ['ie 6', 'ie 7', 'ie 8'].some((e) => {
```
Let's wait until we decide what to do with https://github.com/sonarwhal/sonar/issues/446#issuecomment-324388133 before merging this PR.

qzhou1607 replied:
@molant changes were made based on https://github.com/sonarwhal/sonar/pull/453.

molant replied:
OK! I'll merge this PR once #453 is merged :)

qzhou1607 commented:
Conflicts in `package-lock.json` resolved...
